#**Unstructured Data Analytics (F2021) - Assignment 2**

**Team:**
- Rick Kim
- Gowtami Khambhampati
- Shreya Bakshi
- Iris Wang
- Rathi Kannan
- Alyssa Bae
- Kevin McKinney


**Summary: Building a crowdsourced recommendation system for top 3 beers from beeradvocate.com**
 
https://www.beeradvocate.com/beer/top-rated/

# **Scraper**

In [ ]:
# Use this code if using Colab to run Selenium

# Make sure to go to Runtime -> Change runtime and set GPU as hardware accelerator

# !kill -9 -1 # Use this line to delete this VM and start a new one. 
# The above line deletes all files and folders from the current VM and allocates a new one.

#Selenium is an open-source tool that automates web browsers.
!pip install selenium
!apt-get -q update   #Used to handle installation and removal of softwares and libraries
!apt install -yq chromium-chromedriver #ChromeDriver is a separate executable that Selenium WebDriver uses to control Chrome.
!cp /usr/lib/chromium-browser/chromedriver /usr/bin

import sys
sys.path.insert(0,'/usr/lib/chromium-browser/chromedriver')
from selenium import webdriver 
#WebDriver is a browser automation framework that works with open source APIs. 
#The framework operates by accepting commands, sending those commands to a browser, and interacting with applications.
chrome_options = webdriver.ChromeOptions()
chrome_options.add_argument('--headless')
chrome_options.add_argument('--no-sandbox')
chrome_options.add_argument('--disable-dev-shm-usage')

     |████████████████████████████████| 954 kB 9.0 MB/s 
     |████████████████████████████████| 356 kB 40.3 MB/s 
     |████████████████████████████████| 138 kB 39.4 MB/s 
     |████████████████████████████████| 55 kB 2.3 MB/s 
     |████████████████████████████████| 3.5 MB 34.3 MB/s 
     |████████████████████████████████| 54 kB 2.2 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests 2.23.0 requires urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1, but you have urllib3 1.26.7 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.
Get:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Ign:2 https://developer.down

In [ ]:
# Setting web driver
wd = webdriver.Chrome('chromedriver',options=chrome_options)


In [ ]:
import pandas as pd

# # Variables 
count = 0
comments = pd.DataFrame(columns = ['product_name','product_review','user_rating'])

# # Getting Initial Review Pages 
routes = [] # holds all the review pages 
names = [] # beer names 
route = 'https://www.beeradvocate.com/beer/top-rated/'
wd.get(route)

# # Getting Rows of the 250 Beers Table 
rows = 1 + len(wd.find_elements_by_xpath("//*[@id='ba-content']/table/tbody/tr/td/a"))

print("Getting Beer names")
for r in range(2,220): # Getting 220 Beers 
  element = "//*[@id='ba-content']/table/tbody/tr[{}]/td[2]/a".format(r)
  value = wd.find_element_by_xpath(element).get_attribute("href") 
  element2 = "//*[@id='ba-content']/table/tbody/tr[{}]/td[2]/a/b".format(r)
  names.append(wd.find_element_by_xpath(element2).text)
  
  routes.append(value) # Adding values to the list of links to individual reviews page

print("Loaded Beer Names: ", len(routes))
# Going thru individual reviews page
for idx,route in enumerate(routes): 

  # Fetching routes 
  wd.get(route)

  # Getting ids  
  ids = wd.find_elements_by_xpath("//*[@id='rating_fullview_container']")
   
  # Looping through the page
  print(names[idx],"  ",len(ids))
  for x in range(0, len(ids)):
      # Product Name 
      product_name = names[idx]
      
      # Product Rating 
      user_rating = wd.find_elements_by_xpath('//*[@id="rating_fullview_content_2"]/span[2]')[x].text
      #print(product_rating)

      # Product Review
      product_review = wd.find_elements_by_xpath('//*[@id="rating_fullview_content_2"]')[x].text
      for i in range(1,7): 
        span_elem = '//*[@id="rating_fullview_content_2"]/span[{}]'.format(i)
        if wd.find_elements_by_xpath(span_elem)[x].text is not None:
          span_for_removal = wd.find_elements_by_xpath(span_elem)[x].text
          product_review = product_review.replace(span_for_removal,"")
      product_review = product_review.replace('rDev', "")
      product_review = product_review.replace('\n', " ")
      product_review = product_review.replace('/', "")
      product_review = product_review.replace('look:  | smell:  | taste:  | feel:  | overall: ',"")    
      # Adding data
      comments.loc[len(comments)] = [product_name,product_review,user_rating]
      count +=1 
  # Moving onto the next page
#  print(count)

# Final output
#print(comments)
#print(comments.shape[0])
print('Final: ' + str(count))


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:19: DeprecationWarning: find_element_by_* commands are deprecated. Please use find_element() instead
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:21: DeprecationWarning: find_element_by_* commands are deprecated. Please use find_element() instead


Getting Beer names
Loaded Beer Names:  218


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:33: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:42: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:46: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead


Kentucky Brunch Brand Stout    25


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:49: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:50: DeprecationWarning: find_elements_by_* commands are deprecated. Please use find_elements() instead


Marshmallow Handjee    25
Hunahpu's Imperial Stout - Double Barrel Aged    25
Barrel-Aged Abraxas    25
Heady Topper    25
Pliny The Younger    25
Mornin' Delight    25
King Julius    25
King JJJuliusss    25
SR-71    25
Very Hazy    25
CBS (Canadian Breakfast Stout)    25
Fundamental Observation    25
Pliny The Elder    25
Hunahpu's Imperial Stout - Laird's Apple Brandy Barrel    25
Assassin    25
Fou' Foune    25
Barrel Aged Imperial German Chocolate Cupcake Stout    25
Juice Machine    25
Julius    25
Very Green    25
Zenne Y Frontera    25
Abner    25
Double Sunshine    25
Bourbon County Brand Stout    25
Trappist Westvleteren 12 (XII)    25
Very GGGreennn    25
Dinner    25
KBS    25
Duck Duck Gooze    25
Speedway Stout - Vietnamese Coffee - Bourbon-Barrel Aged    25
Morning Wood    25
Blessed    25
Lou Pepe - Kriek    25
JJJuliusss!    25
Parabola    25
Westly    25
Fuzzy    25
Ann    25
Chemtrailmix    25
Zombie Dust    25
Hunahpu's Imperial Stout    25
Doubleganger    25
Double

In [ ]:
# Saving to CSV

comments.to_csv("beer250.csv", sep="\t")
!ls
from google.colab import files
files.download("beer250.csv")

beer250.csv  sample_data


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# **Task B - Word Frequency Analysis to find top attributes**

In [ ]:
#common imports
import requests
from bs4 import BeautifulSoup
from lxml import html
import pandas as pd
import nltk
from nltk.corpus import stopwords

from string import punctuation
nltk.download('stopwords')
nltk.download('punkt')
from nltk.tokenize import word_tokenize
stop_words = set( stopwords.words('english'))
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import numpy as np
import string

/usr/local/lib/python3.7/dist-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


In [ ]:
# run this cell to upload files.  
from google.colab import files
uploaded = files.upload()

Saving beer250.csv to beer250 (1).csv


In [ ]:
##Reading the beer review file
reviews_df = pd.read_csv(r'beer250.csv',sep='\t', index_col=0)
reviews_df[:20]

product_name  ... user_rating
0   Kentucky Brunch Brand Stout  ...        4.90
1   Kentucky Brunch Brand Stout  ...        5.00
2   Kentucky Brunch Brand Stout  ...        4.56
3   Kentucky Brunch Brand Stout  ...        5.00
4   Kentucky Brunch Brand Stout  ...        5.00
5   Kentucky Brunch Brand Stout  ...        5.00
6   Kentucky Brunch Brand Stout  ...        5.00
7   Kentucky Brunch Brand Stout  ...        4.45
8   Kentucky Brunch Brand Stout  ...        4.15
9   Kentucky Brunch Brand Stout  ...        2.88
10  Kentucky Brunch Brand Stout  ...        5.00
11  Kentucky Brunch Brand Stout  ...        5.00
12  Kentucky Brunch Brand Stout  ...        4.85
13  Kentucky Brunch Brand Stout  ...        4.51
14  Kentucky Brunch Brand Stout  ...        4.42
15  Kentucky Brunch Brand Stout  ...        4.83
16  Kentucky Brunch Brand Stout  ...        5.00
17  Kentucky Brunch Brand Stout  ...        4.90
18  Kentucky Brunch Brand Stout  ...        4.90
19  Kentucky Brunch Brand Stout  ...        4.99

[20 rows x 3 columns]

In [ ]:
reviews_df.groupby(['product_name']).count()

product_review  user_rating
product_name                                                        
A Deal With The Devil                                25           25
A Deal With The Devil - Double Oak-Aged              25           25
Aaron                                                25           25
Abner                                                25           25
Abrasive Ale                                         25           25
...                                                 ...          ...
Westly                                               25           25
Wide Awake It's Morning                              25           25
Zenne Y Frontera                                     25           25
Zombie Dust                                          25           25
§ucaba                                               25           25

[218 rows x 2 columns]

In [ ]:
##Cleaning the review file and removing the stop words
reviews_df = reviews_df.dropna()
reviews_df['cleaned_review'] = reviews_df['product_review'].apply(lambda x :x.translate(str.maketrans('', '', string.punctuation))) # remove punctuation
reviews_df['cleaned_review'] = reviews_df['cleaned_review'].apply(lambda x :x.lower())                                              # make lower
reviews_df['cleaned_review'] = reviews_df['cleaned_review'].apply(word_tokenize).apply(set).apply(list)                    # make a unique list of words

def remove_stopwords(s):
    return [w for w in s if not w in stop_words] 
    
reviews_df['cleaned_review'] =  reviews_df['cleaned_review'].apply(remove_stopwords)

In [ ]:
reviews_df['cleaned_review']


0       [smooth, chocolate, incredible, bottle, syrup,...
1       [smooth, one, hoped, would, id, hype, aroma, w...
2       [feel, else, time, syrup, one, light, goes, mu...
3       [review, bottle, goliath, syrup, tap, difficul...
4       [drinking, time, one, soooo, welltasty, suprem...
                              ...                        
5422    [oak, hint, fofa, clean, taster, glass, tap, e...
5423    [oak, head, light, character, saison, little, ...
5424    [astringent, smooth, glass, lemon, b6, vinous,...
5425    [profile, quite, time, tap, trying, previously...
5426    [feel, apple, lemon, perfect, 6, glowing, tang...
Name: cleaned_review, Length: 5427, dtype: object

In [ ]:
#Creating a combined list of all words for word frequency analysis
all_words = []
for i in range(len(reviews_df)):
    all_words+=reviews_df['cleaned_review'][i]
from nltk import FreqDist
word_freq = nltk.FreqDist(all_words)

In [ ]:
word_freq.most_common()

[('head', 3050),
 ('beer', 2441),
 ('taste', 2414),
 ('pours', 1560),
 ('nose', 1514),
 ('sweet', 1487),
 ('one', 1476),
 ('overall', 1456),
 ('like', 1424),
 ('carbonation', 1424),
 ('dark', 1412),
 ('chocolate', 1379),
 ('finish', 1320),
 ('good', 1319),
 ('mouthfeel', 1307),
 ('aroma', 1280),
 ('nice', 1241),
 ('lacing', 1230),
 ('body', 1228),
 ('well', 1222),
 ('black', 1194),
 ('bottle', 1180),
 ('notes', 1151),
 ('light', 1141),
 ('feel', 1124),
 ('vanilla', 1114),
 ('medium', 1108),
 ('white', 1095),
 ('smell', 1081),
 ('smooth', 1073),
 ('flavor', 1055),
 ('bit', 1046),
 ('great', 1039),
 ('color', 1039),
 ('glass', 1026),
 ('fruit', 1017),
 ('thick', 1005),
 ('really', 993),
 ('bourbon', 967),
 ('coffee', 961),
 ('orange', 958),
 ('poured', 953),
 ('little', 928),
 ('flavors', 916),
 ('bitterness', 874),
 ('brown', 865),
 ('citrus', 860),
 ('barrel', 834),
 ('oak', 795),
 ('sweetness', 771),
 ('creamy', 771),
 ('much', 764),
 ('’', 759),
 ('hazy', 751),
 ('malt', 726),
 ('dry

Top attributes:-


1.  Crisp--Highly carbonated; effervescent
2.  Hoppy--Herbal, earthy, spicy, or citric aromas and flavors of hops
3.  Robust--Rich and full-bodied
4.  fruity --Same as Hoppy
5.  balanced-- Bitterness and Sweetness well represented



In [ ]:
#frequency count of top attributes from reviews
print('crisp',word_freq['crisp'])
print('hoppy',word_freq['hoppy'])
print('robust',word_freq['robust'])
print('fruity',word_freq['fruity'])
print('balanced',word_freq['balanced'])


crisp 219
hoppy 203
robust 75
fruity 355
balanced 635


**Top 3 Attributes selected based on word frequency analysis from customer reviews:**

balanced 635

fruity 355

crisp 219

# **Task C - Perform a similarity analysis using cosine similarity**

In [ ]:
# Upload Attributes.csv file from your local

!rm attributes*.csv
from google.colab import files
uploaded = files.upload()

rm: cannot remove 'attributes*.csv': No such file or directory


Saving attributes.csv to attributes.csv


In [ ]:
#Reading the attributes.csv file
attributes_csv = pd.read_csv("attributes.csv" )
print(attributes_csv)
attributes_csv= attributes_csv['Attributes'].tolist()
print(attributes_csv)

  Attributes
0   balanced
1     fruity
2      crisp
['balanced', 'fruity', 'crisp']


In [ ]:
# choosing only product name and reviews from the original dataframe
col_names = ['product_name','product_review' ]
Taskc_op= reviews_df[col_names]
print(Taskc_op)

                     product_name                                     product_review
0     Kentucky Brunch Brand Stout         2020 vintage drank 102221 Incredible sm...
1     Kentucky Brunch Brand Stout          2020 vintage acquired during the pande...
2     Kentucky Brunch Brand Stout          Long time waiting to tick this one, an...
3     Kentucky Brunch Brand Stout          This review is for the 2019 batch. It ...
4     Kentucky Brunch Brand Stout          Supreme maple OD! Soooo easy drinking ...
...                           ...                                                ...
5422                        Flora          On tap at HF FOFA 2015. HF Taster glas...
5423                        Flora          Batch 4. Appearance is light gold with...
5424                        Flora          B6. Lovely hazy sunshine in the glass....
5425                        Flora          On tap at Festival of Farmhouse Ales 2...
5426                        Flora          Batch 6. Glowing golde

In [ ]:
# cosine similarity function
import numpy as np

def cosine_similarity_calc(vec_1,vec_2):
	cosine = np.dot(vec_1,vec_2)/(np.linalg.norm(vec_1)*np.linalg.norm(vec_2))
	return cosine


In [ ]:
# make an array of tokenized words to vectorize for cosine similarity calculation
reviews_token= reviews_df['cleaned_review']
reviews_token = np.array(reviews_token)
print(reviews_token[0])

['smooth', 'chocolate', 'incredible', 'bottle', 'syrup', 'alcohol', 'glass', 'thick', 'assassin', 'one', 'opened', 'like', 'stouts', 'bourbon', 'soon', 'smell', 'burn', 'vintage', 'breakfast', '102221', 'chased', 'beer', 'tg', 'glad', 'incredibly', '2020', 'still', 'best', 'drank', 'definitely', 'rich', 'maple', 'ton', 'flavor']


In [ ]:
# vectorize tokens and attributes and calculate cosine similarity
cosine_final =[]
for i in range(len(reviews_token)):
  vec_1 = []
  vec_2 = []
  
  #print(reviews_token[0])
  X_set = {w for w in reviews_token[i] if not w in reviews_token}
  Y_set = {w for w in attributes_csv }
  #print("X_Set is " ,X_set)
  #print("Y_Set is " ,Y_set)
  vector = X_set.union(Y_set) 
  for w in vector:
    if w in X_set: 
      vec_1.append(1) 
    else: 
      vec_1.append(0)
    if w in Y_set: 
      vec_2.append(1)
    else: 
      vec_2.append(0)
  cosine = cosine_similarity_calc(vec_1, vec_2)
  #print(cosine)
  cosine_final.append(cosine)
#print(cosine_final)
col3 = pd.DataFrame(cosine_final)
#print(col3)


In [ ]:
# Writing the code in to a csv file

Taskc_op['similarity_score'] = col3
print(Taskc_op)
Taskc_op.to_csv("Taskc_op.csv", index = None)
!ls
from google.colab import files
files.download("Taskc_op.csv")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


                     product_name  ... similarity_score
0     Kentucky Brunch Brand Stout  ...         0.000000
1     Kentucky Brunch Brand Stout  ...         0.000000
2     Kentucky Brunch Brand Stout  ...         0.000000
3     Kentucky Brunch Brand Stout  ...         0.000000
4     Kentucky Brunch Brand Stout  ...         0.000000
...                           ...  ...              ...
5422                        Flora  ...         0.000000
5423                        Flora  ...         0.103695
5424                        Flora  ...         0.109109
5425                        Flora  ...         0.000000
5426                        Flora  ...         0.120386

[5427 rows x 3 columns]
 attributes.csv  'beer250 (1).csv'   beer250.csv   sample_data	 Taskc_op.csv


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
#Reading the output file to a csv
cosine_df = pd.read_csv(r'Taskc_op.csv',sep=',', index_col=0)
cosine_df


product_review  similarity_score
product_name                                                                                    
Kentucky Brunch Brand Stout         2020 vintage drank 102221 Incredible sm...          0.000000
Kentucky Brunch Brand Stout          2020 vintage acquired during the pande...          0.000000
Kentucky Brunch Brand Stout          Long time waiting to tick this one, an...          0.000000
Kentucky Brunch Brand Stout          This review is for the 2019 batch. It ...          0.000000
Kentucky Brunch Brand Stout          Supreme maple OD! Soooo easy drinking ...          0.000000
...                                                                        ...               ...
Flora                                On tap at HF FOFA 2015. HF Taster glas...          0.000000
Flora                                Batch 4. Appearance is light gold with...          0.103695
Flora                                B6. Lovely hazy sunshine in the glass....          0.109109
Flora                                On tap at Festival of Farmhouse Ales 2...          0.000000
Flora                                Batch 6. Glowing golden yellow in colo...          0.120386

[5427 rows x 2 columns]

# **Task D - Sentiment analysis on every product reviews and sorting them (high to low) by the sentiment scores.**

In [ ]:
!pip install vaderSentiment
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
analyser = SentimentIntensityAnalyzer()

     |████████████████████████████████| 125 kB 7.9 MB/s 
     |████████████████████████████████| 127 kB 51.3 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.26.7
    Uninstalling urllib3-1.26.7:
      Successfully uninstalled urllib3-1.26.7
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
selenium 4.0.0 requires urllib3[secure]~=1.26, but you have urllib3 1.25.11 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [ ]:
reviews_df

product_name  ...                                     cleaned_review
0     Kentucky Brunch Brand Stout  ...  [smooth, chocolate, incredible, bottle, syrup,...
1     Kentucky Brunch Brand Stout  ...  [smooth, one, hoped, would, id, hype, aroma, w...
2     Kentucky Brunch Brand Stout  ...  [feel, else, time, syrup, one, light, goes, mu...
3     Kentucky Brunch Brand Stout  ...  [review, bottle, goliath, syrup, tap, difficul...
4     Kentucky Brunch Brand Stout  ...  [drinking, time, one, soooo, welltasty, suprem...
...                           ...  ...                                                ...
5422                        Flora  ...  [oak, hint, fofa, clean, taster, glass, tap, e...
5423                        Flora  ...  [oak, head, light, character, saison, little, ...
5424                        Flora  ...  [astringent, smooth, glass, lemon, b6, vinous,...
5425                        Flora  ...  [profile, quite, time, tap, trying, previously...
5426                        Flora  ...  [feel, apple, lemon, perfect, 6, glowing, tang...

[5427 rows x 4 columns]

In [ ]:
#function to calculate sentiment scores
def sentiment_analyzer_scores(sentence):
    """Generate sentiment score"""
    scores = analyser.polarity_scores(sentence)
    return scores['compound']

#perform sentiment analysis on product review and add to a new column
reviews_df['sentiment'] = reviews_df['product_review'].map(sentiment_analyzer_scores)

In [ ]:
#reviews_df_sentiment =reviews_df.sort_values(by='sentiment', ascending=False)

In [ ]:
## Reviews with the sentiment scores
reviews_df_sentiment=reviews_df
reviews_df_sentiment

product_name  ... sentiment
0     Kentucky Brunch Brand Stout  ...    0.9531
1     Kentucky Brunch Brand Stout  ...    0.5574
2     Kentucky Brunch Brand Stout  ...    0.1159
3     Kentucky Brunch Brand Stout  ...    0.8316
4     Kentucky Brunch Brand Stout  ...    0.9059
...                           ...  ...       ...
5422                        Flora  ...    0.6908
5423                        Flora  ...    0.8910
5424                        Flora  ...    0.9648
5425                        Flora  ...    0.7948
5426                        Flora  ...    0.9260

[5427 rows x 5 columns]

In [ ]:
cosine_df.head(20)

product_review  similarity_score
product_name                                                                                    
Kentucky Brunch Brand Stout         2020 vintage drank 102221 Incredible sm...          0.000000
Kentucky Brunch Brand Stout          2020 vintage acquired during the pande...          0.000000
Kentucky Brunch Brand Stout          Long time waiting to tick this one, an...          0.000000
Kentucky Brunch Brand Stout          This review is for the 2019 batch. It ...          0.000000
Kentucky Brunch Brand Stout          Supreme maple OD! Soooo easy drinking ...          0.000000
Kentucky Brunch Brand Stout          I have now had 4 different years of KB...          0.000000
Kentucky Brunch Brand Stout          2020 Bottle.  Absolutely bonkers Maple...          0.000000
Kentucky Brunch Brand Stout          Some burnt sugar and a bit of licorice...          0.000000
Kentucky Brunch Brand Stout          Bottle @ Presidential Suite. Thanks, A...          0.000000
Kentucky Brunch Brand Stout          The look is great, and then it's downh...          0.000000
Kentucky Brunch Brand Stout          This dark nectar poured like a heavenl...          0.000000
Kentucky Brunch Brand Stout          Is this the best beer in the world, ma...          0.103695
Kentucky Brunch Brand Stout          I had this for the first time in 2020....          0.000000
Kentucky Brunch Brand Stout          Rich dark brown with a thin finger of ...          0.000000
Kentucky Brunch Brand Stout          Poured at cellar temp into a teku. Bot...          0.000000
Kentucky Brunch Brand Stout          2020 recent release Bottle courtesy of...          0.000000
Kentucky Brunch Brand Stout          This review is for the 2020 KBBS from ...          0.000000
Kentucky Brunch Brand Stout         2020 on draft at Toppling Goliath.  It ...          0.000000
Kentucky Brunch Brand Stout         2019 KBBS Look: black syrup, no head, s...          0.092450
Kentucky Brunch Brand Stout          2019 draft and bottle.  I didn't think...          0.079305

In [ ]:
reviews_df_sentiment.head(20)

product_name  ... sentiment
0   Kentucky Brunch Brand Stout  ...    0.9531
1   Kentucky Brunch Brand Stout  ...    0.5574
2   Kentucky Brunch Brand Stout  ...    0.1159
3   Kentucky Brunch Brand Stout  ...    0.8316
4   Kentucky Brunch Brand Stout  ...    0.9059
5   Kentucky Brunch Brand Stout  ...    0.9022
6   Kentucky Brunch Brand Stout  ...   -0.6114
7   Kentucky Brunch Brand Stout  ...    0.9274
8   Kentucky Brunch Brand Stout  ...    0.9442
9   Kentucky Brunch Brand Stout  ...    0.5994
10  Kentucky Brunch Brand Stout  ...    0.9241
11  Kentucky Brunch Brand Stout  ...    0.9300
12  Kentucky Brunch Brand Stout  ...    0.6341
13  Kentucky Brunch Brand Stout  ...    0.9892
14  Kentucky Brunch Brand Stout  ...    0.9534
15  Kentucky Brunch Brand Stout  ...    0.4329
16  Kentucky Brunch Brand Stout  ...    0.9864
17  Kentucky Brunch Brand Stout  ...    0.9655
18  Kentucky Brunch Brand Stout  ...    0.7813
19  Kentucky Brunch Brand Stout  ...    0.9550

[20 rows x 5 columns]

In [ ]:
## Merging the dataframes to have both sentiment and similarity scores
review_sent_similar= pd.merge(reviews_df_sentiment, cosine_df, on='product_review')

In [ ]:
review_sent_similar.head(20)

product_name  ... similarity_score
0   Kentucky Brunch Brand Stout  ...         0.000000
1   Kentucky Brunch Brand Stout  ...         0.000000
2   Kentucky Brunch Brand Stout  ...         0.000000
3   Kentucky Brunch Brand Stout  ...         0.000000
4   Kentucky Brunch Brand Stout  ...         0.000000
5   Kentucky Brunch Brand Stout  ...         0.000000
6   Kentucky Brunch Brand Stout  ...         0.000000
7   Kentucky Brunch Brand Stout  ...         0.000000
8   Kentucky Brunch Brand Stout  ...         0.000000
9   Kentucky Brunch Brand Stout  ...         0.000000
10  Kentucky Brunch Brand Stout  ...         0.000000
11  Kentucky Brunch Brand Stout  ...         0.103695
12  Kentucky Brunch Brand Stout  ...         0.000000
13  Kentucky Brunch Brand Stout  ...         0.000000
14  Kentucky Brunch Brand Stout  ...         0.000000
15  Kentucky Brunch Brand Stout  ...         0.000000
16  Kentucky Brunch Brand Stout  ...         0.000000
17  Kentucky Brunch Brand Stout  ...         0.000000
18  Kentucky Brunch Brand Stout  ...         0.092450
19  Kentucky Brunch Brand Stout  ...         0.079305

[20 rows x 6 columns]

In [ ]:
#Sorting the dataframe based on sentiment score
review_sent_similar =review_sent_similar.sort_values(by='sentiment', ascending=False)

In [ ]:
review_sent_similar

product_name  ... similarity_score
164           Mornin' Delight  ...         0.000000
4587                  Abraxas  ...         0.000000
2925          Cable Car Kriek  ...         0.076640
3540                 No Rules  ...         0.000000
2794                The Peach  ...         0.000000
...                       ...  ...              ...
230                     SR-71  ...         0.000000
3157            Scaled Way Up  ...         0.000000
591           Double Sunshine  ...         0.035202
4848                     Matt  ...         0.000000
3443  Kaggen! Stormaktsporter  ...         0.000000

[5429 rows x 6 columns]

# **Task E -  Top 3 beer recommendations based on similarity and sentiment scores**

In [ ]:
reviews_combined=review_sent_similar.groupby('product_name')[['similarity_score','sentiment']].mean()

In [ ]:
reviews_combined['recommend'] = reviews_combined['similarity_score']+reviews_combined['sentiment']

In [ ]:
reviews_combined

similarity_score  sentiment  recommend
product_name                                                                   
A Deal With The Devil                            0.011350   0.719492   0.730842
A Deal With The Devil - Double Oak-Aged          0.008007   0.844124   0.852131
Aaron                                            0.006975   0.821640   0.828615
Abner                                            0.049218   0.855920   0.905138
Abrasive Ale                                     0.020330   0.729304   0.749634
...                                                   ...        ...        ...
Westly                                           0.033919   0.724272   0.758191
Wide Awake It's Morning                          0.020756   0.655404   0.676160
Zenne Y Frontera                                 0.043571   0.880060   0.923631
Zombie Dust                                      0.039927   0.816032   0.855959
§ucaba                                           0.024871   0.838016   0.862887

[218 rows x 3 columns]

In [ ]:
##Recommendations based only on similarity & sentiment scores
reviews_combined.sort_values(by='recommend', ascending=False)[0:3]

similarity_score  sentiment  recommend
product_name                                           
Cable Car Kriek          0.032246   0.911180   0.943426
Cable Car                0.024942   0.904356   0.929298
Flora Plum               0.035451   0.891708   0.927159

Beer Preferences are extremely subjective and we have used the entire dataset consisting of all the reviews instead of top 100 or top 200 reviews based on similarity score. Hence products with low similarity scores but having high sentiments are also getting recommended.

# **Task F - Using spaCy package to determine similarity and provide top-3 beer recommendation**

In [ ]:
#common imports for spacy
!python -m spacy download en_core_web_lg

     |████████████████████████████████| 827.9 MB 2.3 MB/s 
  Created wheel for en-core-web-lg: filename=en_core_web_lg-2.2.5-py3-none-any.whl size=829180942 sha256=1edb5aa619cdc4b41d5c398aeba8e980681f370370f98d86919ba6b47026f05d
  Stored in directory: /tmp/pip-ephem-wheel-cache-b5rfh9ag/wheels/11/95/ba/2c36cc368c0bd339b44a791c2c1881a1fb714b78c29a4cb8f5
Successfully built en-core-web-lg
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_lg')


In [ ]:
import spacy
import en_core_web_lg

nlp = en_core_web_lg.load()
  
#Create function to join tokenized words into a sentence 
def join_words(comment):
  return " ".join(comment)

#Map function on product reviews
reviews_df['joined_review'] = reviews_df['cleaned_review'].map(join_words)

In [ ]:
#Function to calculate similarity with pre-processing functions
def calculate_similarity(comment):
 
  base = nlp(comment)
  compare = nlp(input_attributes)
  return base.similarity(compare)

In [ ]:
input_list = ['balanced', 'fruity', 'crisp']
input_attributes = " ".join(input_list)
reviews_df['similarity'] = reviews_df['joined_review'].map(calculate_similarity)


In [ ]:
# recommendations based on spacy similarity
reviews_df.sort_values(by='similarity', ascending=False)[0:3]

product_name  ... similarity
2175  Galaxy Dry Hopped Fort Point Pale Ale  ...   0.862940
4194                         Saison Bernice  ...   0.858119
3608                      Lou Pepe - Gueuze  ...   0.842679

[3 rows x 7 columns]

In [ ]:
# recommendations based on spacy similarity and sentiment
reviews_df.sort_values(['sentiment','similarity'], ascending=[False, False])[0:3]

product_name  ... similarity
164   Mornin' Delight  ...   0.684986
4585          Abraxas  ...   0.706428
2923  Cable Car Kriek  ...   0.700020

[3 rows x 7 columns]

The top-3 beers based on word vector approach is much different than the bag of words. Again, the sentiment scores and similarity numbers are not always correlated. As shown in cell 41, purely based off of similarity scores and disregarding sentiment,Saison Bernice will be in the top-3. The sentiment score for Saison Bernice clearly shows a negative sentiment. Therefore, recommendations for beers in either approaches should be accompanied by sentiment evaluation.

# **Task G**